### PR TITLE
Add tutorial for running OpenSAFELY via Gitpod 

### PR DESCRIPTION
**Pull request by StevenMaude**
## What this PR does

* It describes the use of Gitpod for the OpenSAFELY template repository in the Getting Started guide via the use of tabs.
* Gitpod becomes the default path in the guide, unless someone deliberately selects the Windows details in the guide. The Windows details formed the basis of these changes, and are still included in those tabs.

This is an initial iteration. We aim to build quickly and improve this guide to fill any explanation gaps. @brianmackenna has run through this guide in a pairing session. We plan to run similar sessions to get feedback from more people.

## Background

In May, @bloodearnest and I tested [Gitpod](https://gitpod.io). Gitpod is a viable option to run OpenSAFELY tooling without a user having to install any extra software on their own computer.

Gitpod operates in modern web browsers cross-platform. Gitpod provides an online Visual Studio Code editor and virtual machine for running code, and allows users to push changes to GitHub. This should reduce the technical barriers for onboarding users. Gitpod currently has a generous free usage limit for public repositories, a month's trial for private repositories and may be relatively inexpensive for educational — and perhaps research? — users.

Since Gitpod uses Visual Studio Code and Visual Studio Code is available cross-platform and without cost, this also suggests that Visual Studio Code might be a default development environment that we propose to new users. Experienced OpenSAFELY users or people with more software development experience might still choose their own tooling.

### Inclusion of multiple tutorial routes (likely imminent, if we decide on this)

After a initial run through the tutorial with @brianmackenna, we discussed whether it might be better having separate pages for the guides after all. This might work via using Python `markdown-include` with mkdocs, so that we can include the duplicated content in multiple places without actually copy-pasting.

This could be a subsequent iteration, as this will require some considerable restructuring.

### Use of GitHub Codespaces (longer term)

If GitHub Codespaces eventually becomes publicly available, that might be an alternative in future. GitHub Codespaces looks very similar in features to Gitpod and also uses Visual Studio Code.

### CODE REVIEW

File: docs/gitpod.md
Comment (StevenMaude) [reply docs/gitpod.md:107]: :+1:

File: docs/gitpod.md
Comment (StevenMaude) [reply docs/gitpod.md:142]: Yes. The way this is written is not how it should be eventually, if we incorporate it.

The tabs display is not something I recall seeing; that's a nice way of structuring it. We could have a setup and installtion section that is tabbed for each platform. Then the main document covers the OpenSAFELY instructions.

File: docs/gitpod.md
Comment (StevenMaude) [reply docs/gitpod.md:295]: We don't need to at all; I just documented it here, because it didn't work out of the box, and if anyone wanted to follow along to try things out.

File: docs/gitpod.md
Comment (StevenMaude) [reply docs/gitpod.md:37]: Yep, fine! In practice, I wouldn't expect someone just investigating OpenSAFELY to encounter the current usage limit, and those users are where this is potentially a better fit right now. 

People working on it more regularly might still favour configuring their own machine. Or they can pay a small amount per month, or self-hosting for OpenSAFELY users might be an option.

**Comment by StevenMaude** [comment]
It's possible to configure a Gitpod workspace on launch (and optionally to speed things up by readying the workspace on changes to a `.gitpod.yml`), so it's possible to ditch the installation of `opensafely`.

Not quite sure about getting `jupyter lab` running. I get an error if I naively try to start it as a `command` in `.gitpod.yml`. But the command then works fine when entered manually when the workspace starts; the thing being run has the same path (which is in `…/.pyenv/shims`).

**Comment by StevenMaude** [comment]
I'm starting to integrate the document I first wrote here with the existing "Getting Started" guide, using tabs for alternate content as suggested by @sebbacon. This isn't necessarily how it will ultimately be, but is a first attempt.

This approach does have the advantage that it might eventually unify all installation instructions into one location. That would solve #117: that issue is out of scope for this PR.

I'm currently migrating the new information from the document I added and into the "Getting Started" guide. Any other niceties like configuring a workspace on clone are not yet discussed. `opensafely` is still manually installed with `pip`. They may be introduced later if we incorporate a suitable `.gitpod.yml` into the research template.

Tasks:
- [ ] First pass merge of new Gitpod content into the "Getting Started guide" (in progress, aim to finish next day I'm around at DataLab).
- [ ] Ask for feedback from others on this format.
- [ ] Add some screenshots of Gitpod.
- [ ] Develop a Gitpod configuration for the research template to further simplify the instructions.
- [ ] Address any remaining TODOs in the new text.

**Comment by StevenMaude** [comment]
This is still missing screenshots, but the Gitpod instructions are now part of the Getting Started guide. Tabbed sections split between Windows and Gitpod instructions, although different tabbed sections are independent.

We'd need some JavaScript on the page to toggle all instructions.

Alternatively, there is a recently added [sponsor-only feature for this in the Material for Mkdocs theme](https://github.com/squidfunk/mkdocs-material/issues/2678). This may be in the public release in future.

**Comment by StevenMaude** [comment]
I've tidied this up more.

All that's left from the [suggested tasks](https://github.com/opensafely/documentation/pull/249#issuecomment-849805150) for a first pass of the documentation is to add Gitpod screenshots.

We then should review this new format, decide if we want it, or we need to split out the guides separately. The Gitpod content should also be reviewed and improved.

**Review by sebbacon — COMMENTED** [review]
Looks great, I think it works well.

My suggestions are mainly around pruning stuff. My preference is to err towards brevity rather than completeness and/or use pictures to make it feel less work to follow.

File: docs/getting-started.md
Comment (StevenMaude) [reply docs/getting-started.md:223]: There aren't any serious consequences. I agree that it distracts from the actual task at hand.

I don't know how obvious Gitpod is to other people. I know that how workspaces operate wasn't immediately obvious, because you don't see the list of workspaces when you open one up. We could move this inside the Reference section, or potentially drop it if it's not useful.

File: docs/getting-started.md
Comment (StevenMaude) [reply docs/getting-started.md:393]: Fine :slightly_smiling_face.

The reason for including was primarily as a way to restore the options if you've lost them; I've had this before in software when accidentally closing a toolbar and it's infuriating.

File: docs/getting-started.md
Comment (StevenMaude) [reply docs/getting-started.md]: Yes, agree and applies to much of the other text added, I'm including lots of screenshots :framed_picture:

File: docs/getting-started.md
Comment (StevenMaude) [reply docs/getting-started.md:89]: I'll try changing it.

I find the end of tabbed sections difficult to work out at times. This is a bit of a problem with the theme we're using. Typically I've added headings after tabbed sections to try to demarcate it more.

The other option is that we override the theme CSS and do something like stick another line underneath.

Comment (sebbacon) [reply docs/getting-started.md:89]: Yes. I'm in favour of a light CSS override, it probably makes sense. Though usually I see this in the concept of *code* snippets like this - it may look a bit funny for narrative snippets

image

Comment (sebbacon) [reply docs/getting-started.md:89]: Could you try something ugly like a coloured background using browser DOM manipulation and paste the results here for more discussion? That'd be good enough to have a sense of if it'd work, and then we could ask Tom to do something nice if we like it?

File: docs/getting-started.md
Comment (sebbacon) [reply docs/getting-started.md:393]: I'm quite up for a "more about gitpod" page or section where we can add tips, if we feel that's useful - just as long as we keep most tips out of the main flow

**Comment by sebbacon** [comment]
On sticky tabs: I feel we make enough use of mkdocs to warrant $10 / month, but of course there's the challenge of getting it through the finance process; I'd love to get this released today if we can. Perhaps the maintainer would be happy to give us access on trust based on our uni brand.

**Comment by sebbacon** [comment]
Per [conversation in Slack this morning](https://ebmdatalab.slack.com/archives/C63UXGB8E/p1623217496140100?thread_ts=1623145544.089100&cid=C63UXGB8E), should we add a section (for now) about "for the time being, you'll still want to get all this running on your own computer if you're going to do a real project"?

**Comment by StevenMaude** [comment]
> On sticky tabs: I feel we make enough use of mkdocs to warrant $10 / month, but of course there's the challenge of getting it through the finance process; I'd love to get this released today if we can. Perhaps the maintainer would be happy to give us access on trust based on our uni brand.

Paying seems reasonable; it's more I'm not sure the implications of that for local development or anyone external wanting to build the docs (which might be zero people). Not sure how using the premium theme works as a paid dependency, and I didn't look much into it.

None of this is immediately important, just nice to fix/polish.

**Comment by StevenMaude** [comment]
The current `study_definition.py` has:

```python
from cohortextractor import StudyDefinition, patients, codelist, codelist_from_csv
```

which gets complaints with the Python extension — `codelist`, `codelist_from_csv` are unused — and the line's not PEP8 compliant :upside_down_face:

I'm assuming I can just take the latter two imports out, which will fix everything.

(There's definitely a case for testing/linting the code in docs, if that's possible.)

**Comment by bloodearnest** [comment]
> The current `study_definition.py` has:
> 
> ```python
> from cohortextractor import StudyDefinition, patients, codelist, codelist_from_csv
> ```
> 
> which gets complaints with the Python extension that we use in VSCode via Gitpod and the research template — `codelist`, `codelist_from_csv` are unused — and the line's not PEP8 compliant 
> 
> I'm assuming I can just take the latter two imports out, which will fix everything.
> 
> (There's definitely a case for testing/linting the code in docs, if that's possible.)

For the research-template, I added `# noqa` to the end of this line, as it couldn't find `cohortextractor` package in the project Which is understandable, as it's run inside docker.

That was a pylint error though, so I switched to flake8 as the default linter, which maybe doesn't complain about that.

**Comment by StevenMaude** [comment]
Related irritation: the flake8 check in VSCode complains there is no newline at the end of files (`report.py`, `study_definition.py`), when copy-pasted from examples. I'm not even sure if we add a newline at the end that a reader would copy-paste the empty line.

File: docs/getting-started.md
Comment (StevenMaude) [reply docs/getting-started.md]: Done.

File: docs/getting-started.md
Comment (StevenMaude) [reply docs/getting-started.md]: Removed!

File: docs/gitpod.md
Comment (StevenMaude) [reply docs/gitpod.md]: Think this is out of scope for this PR; it's in the [list of tasks](https://github.com/opensafely/documentation/pull/249#issuecomment-849805150) below and I'll move the "post-first-iteration" tasks to a separate issue when this PR is merged.

File: docs/getting-started.md
Comment (StevenMaude) [reply docs/getting-started.md]: Done.

File: docs/getting-started.md
Comment (StevenMaude) [reply docs/getting-started.md:89]: tabbed-css-override-background-color

Comment (StevenMaude) [reply docs/getting-started.md:89]: Going to mark as resolved for now, because we may move away from tabs completely. Otherwise, we need to revisit it.

**Review by bloodearnest — APPROVED** [review]

**Comment by bloodearnest** [comment]
This has a had a lot of feedback, and some user testing, so I'm going to land in @StevenMaude's abcense.

**Pull request merged**
